packages/SwingSet/src/liveslots/watchedPromises.js

```diff
@@ -0,0 +1,162 @@
+// @ts-check
+import { assert } from '@agoric/assert';
+import { M } from '@agoric/store';
+import { parseVatSlot } from '../lib/parseVatSlots.js';
+
+export function makeWatchedPromiseManager(
+  syscall,
+  vrm,
+  vom,
+  cm,
+  convertValToSlot,
+  convertSlotToVal,
+) {
+  const { makeScalarBigMapStore } = cm;
+  const { defineDurableKind } = vom;
+
+  // watched promises by vpid: each entry is an array of watches on the
+  // corresponding vpid; each of these is in turn an array of a watcher object
+  // and the arguments associated with it by `watchPromise`.
+  let watchedPromiseTable;
+
+  function pseudoThen(p, vpid) {
+    function settle(value, wasFulfilled) {
+      const watches = watchedPromiseTable.get(vpid);
+      watchedPromiseTable.delete(vpid);
+      for (const watch of watches) {
+        const [watcher, ...args] = watch;
+        const handler = wasFulfilled ? watcher.onFulfilled : watcher.onRejected;
+        if (handler) {
+          handler(value, ...args);
+        }
+      }
+    }
+
+    p.then(
+      res => settle(res, true),
+      rej => settle(rej, false),
+    );
```

Comment:
Yes and no.  It does rely on the promise being in RAM but if there's an imported promise it's going to be in RAM anyway, so doing this costs us nothing extra.  Keep in mind that this is not a mechanism to virtualize promises, it's more a way to virtualize `then` calls.  In this case virtualization is not for the purpose of reducing the RAM footprint but for the purpose of gaining durability.  In other words, if instead of using a `promiseWatcher` someone had done a regular `then` call they'd also be holding it in RAM (albeit their attention to the promise would be volatile) -- which is what you'd want because GC of the promise would otherwise cause it to be released by the vat and the import dropped, and so the kernel would not send a notification to the vat when the promise is resolved.

Once we have actual virtual promises then we'll be in a position to keep an on-disk refcount for a promise, keeping the promise alive from the kernel's perspective but allowing the JavaScript promise object to disappear from RAM inside the vat, and then having a mechanism whereby the delivery of the promise resolution notification from the kernel would trigger the flow into whatever handlers had been registered.  But this is not that.

Reply:
> if there's an imported promise it's going to be in RAM anyway

I didn't realize that the promise instance created by an imported promise was held strongly by anything in liveslots.

> In this case virtualization is not for the purpose of reducing the RAM footprint but for the purpose of gaining durability.

Ack, was just hoping this could put us on the path to both.

> having a mechanism whereby the delivery of the promise resolution notification from the kernel would trigger the flow into whatever handlers had been registered

Right, that's indeed what would be needed.